compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECanonicalizeNpuDmaCpyNd.cpp

```diff
@@ -0,0 +1,187 @@
+      SmallVector<OpFoldResult> srcStrides = dmaOp.getSourceMixedStrides();
+
+      SmallVector<OpFoldResult> tgtOffsets = dmaOp.getTargetMixedOffsets();
+      SmallVector<OpFoldResult> tgtSizes = dmaOp.getTargetMixedSizes();
+      SmallVector<OpFoldResult> tgtStrides = dmaOp.getTargetMixedStrides();
+
+      bool allRankLessThan4 = srcOffsets.size() <= 4 && srcSizes.size() <= 4 &&
+                              srcStrides.size() <= 4 &&
+                              tgtOffsets.size() <= 4 && tgtSizes.size() <= 4 &&
+                              tgtStrides.size() <= 4;
+      if (!allRankLessThan4) {
+        dmaOp.emitOpError() << " has offsets/sizes/strides with more than 4 "
+                               "elements, unexpected in this pass.";
+        return WalkResult::interrupt();
+      }
+
+      if (dmaOp.getSourceMemorySpaceAsUInt() == 0) {
+        if (!dmaOp.hasSourceAddressing()) {
+          dmaOp.emitOpError()
+              << "has source in L3, but does not have source addressing. "
+                 "Source addressing is required to canonicalize to rank-4 "
+                 "here.";
+          return WalkResult::interrupt();
+        }
+        // Make rank-4.
+        srcOffsets = getPrepended(srcOffsets, zero);
+        srcSizes = getPrepended(srcSizes, one);
+        srcStrides = getPrepended(srcStrides, zero);
+
+        // Ensure that the zero stride is in the outer-most dimension.
+        std::optional<uint32_t> maybeSwapIndex =
+            getZeroStrideIndex(srcSizes, srcStrides, dmaOp);
+        if (!maybeSwapIndex.has_value()) {
+          return WalkResult::interrupt();
+        }
+        uint32_t swapIndex = maybeSwapIndex.value();
+        std::swap(srcOffsets[0], srcOffsets[swapIndex]);
+        std::swap(srcSizes[0], srcSizes[swapIndex]);
+        std::swap(srcStrides[0], srcStrides[swapIndex]);
+      }
+
+      if (dmaOp.getTargetMemorySpaceAsUInt() == 0) {
+        if (!dmaOp.hasTargetAddressing()) {
+          dmaOp.emitOpError()
+              << "has target in L3, but does not have target addressing. "
+                 "Target addressing is required to canonicalize to rank-4 "
+                 "here.";
+          return WalkResult::interrupt();
+        }
+
+        // Make rank-4.
+        tgtOffsets = getPrepended(tgtOffsets, zero);
+        tgtSizes = getPrepended(tgtSizes, one);
+        tgtStrides = getPrepended(tgtStrides, zero);
+
+        // Ensure that the zero stride is in the outer-most dimension.
+        std::optional<uint32_t> maybeSwapIndex =
+            getZeroStrideIndex(tgtSizes, tgtStrides, dmaOp);
+        if (!maybeSwapIndex.has_value()) {
+          return WalkResult::interrupt();
+        }
+        uint32_t swapIndex = maybeSwapIndex.value();
+        std::swap(tgtOffsets[0], tgtOffsets[swapIndex]);
+        std::swap(tgtSizes[0], tgtSizes[swapIndex]);
+        std::swap(tgtStrides[0], tgtStrides[swapIndex]);
```

Comment:
Swap would be the easiest and time-complexity efficient, but shouldn't we want a stable-sort equivalent here? i.e. preserving the order of rest of the non-zero stride dimensions?

Reply:
I think I was trying to be as close to the original (see  [canonicalizeNpuStridedPatternForAIE](https://github.com/nod-ai/iree-amd-aie/blob/f8f31a8100af2f3c464abfc868c003afbc7f8dbb/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp#L557)) but what you describe is more intuitive, and will be correct too. I'll use [c++ rotate](https://en.cppreference.com/w/cpp/algorithm/rotate)